[MRG]: BUG: fix MPI child communication, add several improvements w/ unit testing

blakecaldwell opened the pull request:
This should address the issues with #190, finally (I've tested on a rebased branch with code from #190). The key fix was simple: 2e425c53ad60fe223ede18050a52d047c2202447. When the system is slow (e.g. Travis) is might accumulate more than 4k in the stderr buffer and only the first 4k would get read.

During the search for this bug, I made many improvements that should make life easier for everyone going forward. The changes to the MPI child communication scheme have a bunch of unit tests. The largest change for discussion is the testing scheme code for parallel backends... @jasmainak @rythorpe @ntolley 

It drove me mad trying to fix Travis issues when the time to complete tests and print errors is ~10 minutes. While we are still validating the MPIBackend and adding new simuatlion outputs, I think having the full simulation is useful. So I updated `test_parallel_backends.py` to make my life easier, and it's included in this PR.

1. There is a custom pytest marker `incremental` that is defined in `conftest.py` and declared in `pytest.ini`. It allows a failing test (e.g. a reduced mpi simulation) to cancel subsequent tests (e.g. the full lengthl mpi and joblib comparison test). This was an example from pytest's documentation and it doesn't require any additional modules.
2. If `mpi4py` isn't installed, then MPI tests are skipped, but that does cause subsequent tests to be skipped (as long as they don't require `mpi4py`). This seems to be the appropriate solution to the issue @ntolley ran into when testing locally without `mpi4py`.
3. The test `test_mpi_nprocs` is pretty important to make a prerequisite for later mpi tests. If there is truly only one processor available, then mpi tests aren't useful tests (and the user should be aware).
4. Each test in `test_parallel_backends.py` performs a specific purpose. Code for running simulations is made as modular as possible. I find this easier to follow than the previous `test_compare_hnn` that ran reduced and full-length simulations all together.

blakecaldwell commented:
Ugh, I need to redo this yet again.

jasmainak commented:
okay, let us know when ready!

jasmainak commented:
I took a brief look and it looks good from a superficial look. I'd like to set aside some time and mess around with the code / review it more thoroughly though. Might push a commit or two. Just to absorb what's going on. This is turning out to be quite the saga. Thanks for your efforts @blakecaldwell

blakecaldwell commented on hnn_core/parallel_backends.py:
```diff
@@ -340,9 +385,6 @@ def simulate(self, net):
         os.close(pipe_stdin_r)
 
-        # data will be stored here; output will be printed and discarded
-        self.proc_data_bytes = b''
```
Moved to `__init__` so that tests can use this attribute.

blakecaldwell commented on hnn_core/parallel_backends.py:
```diff
@@ -351,8 +393,23 @@ def simulate(self, net):
+        timeout = 0
         # loop while the process is running
-        while proc.poll() is None:
+        while True:
```
A clean break is when: (1) the child process has sent the completion signal and (2) the child process has exited (i.e. `proc.poll() is not None`

blakecaldwell commented on hnn_core/parallel_backends.py:
```diff
@@ -364,17 +421,17 @@ def simulate(self, net):
-                            raise ValueError("Invalid data send completion "
-                                             "signal from child MPI process")
+                                          self._read_stderr)
+                    elif isinstance(completion_signal, int):
```
The value returned from `_read_stderr` is just an integer now.

blakecaldwell commented:
> I took a brief look and it looks good from a superficial look. I'd like to set aside some time and mess around with the code / review it more thoroughly though. Might push a commit or two. Just to absorb what's going on. This is turning out to be quite the saga. Thanks for your efforts @blakecaldwell

Sure, would recommend starting with my review comments to see the important changes. Also, the addition of `conftest.py`.

blakecaldwell commented on hnn_core/mpi_child.py:
```diff
@@ -62,34 +62,37 @@ def _read_params(self):
 
-        # the parent process is waiting for "end_of_sim:[#bytes]" with the
-        # length of data
-        sys.stdout.write('end_of_data:%d' % len(pickled_bytes))
-        sys.stdout.flush()  # flush to ensure signal is not buffered
+        # the parent process is waiting for "@end_of_sim:[#bytes]@" with the
+        # length of data. The '@' is not found in base64 encoding, so we can
+        # be certain it is the border of the signal
+        sys.stderr.write('@end_of_data:%d@' % len(pickled_bytes))
+        sys.stderr.flush()  # flush to ensure signal is not buffered
```
The signal is written to `stderr` (along with the data) instead of `stdout`. The `.flush()` calls ensure sequential ordering of (1) pickled data and (2) 'end_of_data` signal. I can't think of a scenario (3) would be more pickled data, but the code in `_read_stderr` would still work if that were the case.

blakecaldwell commented on hnn_core/tests/conftest.py:
```diff
@@ -0,0 +1,59 @@
+""" Example from pytest documentation
+
+https://pytest.org/en/stable/example/simple.html#incremental-testing-test-steps
+"""
```
This example did (almost) exactly what I wanted.

blakecaldwell commented on hnn_core/tests/conftest.py:
```diff
@@ -0,0 +1,59 @@
+
+    if "incremental" in item.keywords:
+        # incremental marker is used
+        if call.excinfo is not None and not call.excinfo.typename == "Skipped":
```
I added `and not call.excinfo.typename == "Skipped"` so that skipped tests (i.e. `mpi4py` not being installed) would not cause subsequent tests in the incremental block to be xfailed.

blakecaldwell replied:
> humm ... how should we interpret xfailed? Will it cause CI to fail?
> 
It's the same as failed. CI should fail if MPI tests fail, no?

> also, can you add a comment somewhere indicating that this is the only line you modified and everything else is copied. Will make it easier to maintain the code.

No problem

blakecaldwell replied:
Note that xfailed in this case will only follow a true failure. CI will fail.

blakecaldwell commented on hnn_core/tests/test_mpi_child.py:
```diff
@@ -9,8 +12,70 @@
+                       "process did not contain data length."):
+        backend._read_stderr(pipe_stderr_r, selectors.EVENT_READ)
+
+    stderr.write("blahblah@end_of_data:1000@blah")
```
The `blah` at the end would not be anticipated irl as mentioned in an earlier comment. Though, it helps to test the separation of the signal from the data in `_read_stderr()`

jasmainak replied:
maybe we're missing the test case of `blahblah@end_of_data:100` ? This should fail, right?

blakecaldwell replied:
That seems like a good test to add.

rythorpe commented:
Apologies if I've been a bit absent lately @blakecaldwell. I'll try to get caught up with this PR this weekend.

jasmainak commented on hnn_core/parallel_backends.py:
```diff
@@ -351,8 +393,23 @@ def simulate(self, net):
         # loop while the process is running
-        while proc.poll() is None:
+        while True:
+            if not proc.poll() is None:
+                if completed is True:
+                    break
+                elif timeout > 4:
+                    # This is indicative of a failure. For debugging purposes.
+                    warn("Timed out (5s) waiting for end of data after child "
+                         "process stopped")
+                    break
+                else:
+                    timeout += 1
+                    sleep(1)
```
isn't this part of some system library? I haven't looked deeply but it feels like [Popen.wait](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait) might do the same?

blakecaldwell replied:
The problem with calls like `Popen.wait()` is that they block. In this case, the checks are for loop control... and the combination of conditions wouldn't match a typical system library. I'm making two specific checks if the loop should be continued and neither should block.

blakecaldwell replied:
> very dumb question, what's the reason for encoding the data as base64? I just googled and found this. One of the answers says: "Base64 is just an inefficient method to store binary data." I trust you on this, I am just curious what is the reason you pickle and then encode the data as base 64? What are the pros/cons of different choices (if any)?

No, good question. Starting with the limitations: stdout and stderr can only pass strings. Pikle returns a string, except it's in the 'latin-1' encoding. Base64 is a way to convert the byte string into an ASCII encoding (i.e. portable). I remember finding support for using this method on stackoverflow when sending pickled data over a network and particularly the `cloudpickle` project (same `Popen` use case as us). However, I rechecked that project and they are no longer using baseX encoding! It looks like the reason for base32/base64 was for python2/windows compatibility. I'm interpreting python3 with windows as being okay.

https://github.com/cloudpipe/cloudpickle/commit/65b240866c7d861e79fcc68dc68ce06f3e1d6f57#diff-9d24dc3dc9330ef153577f547b4024b50a095ce91290796e44b7a59dcd7b9119L49-R105

Straight up pickle loads and dumps should work!

jasmainak replied:
> The problem with calls like Popen.wait() is that they block.

So just to complete my understanding, you don't want them to block because of the GUI?

blakecaldwell replied:
> > The problem with calls like Popen.wait() is that they block.
> 
> So just to complete my understanding, you don't want them to block because of the GUI?

Even, with the command-line. If a `wait()` call is made, then simulation progress isn't printed to the screen until `wait()` returns. Same reason as for the GUI.

blakecaldwell replied:
> Straight up pickle loads and dumps should work!

Unfortunately, not the case. I tried with several different ways, but could not successfully unpickle the data without using base64 encoding. Here's a similar issue:
https://stackoverflow.com/questions/30469575/how-to-pickle-and-unpickle-to-portable-string-in-python-3

I will keep using base64 for this PR

jasmainak reviewed (commented):
Looks very promising! Code gets better and better with each PR.

So, I did not get through the entire PR. I still need to do `test_mpi_backends.py`. Please interpret all the flurry of comments as my excitement to see interesting progress, specially with the testing procedure!

One last question, wondering how all this compares to [Popen.communicate](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate)? If there are system library alternatives, we should try to use them.

blakecaldwell commented on hnn_core/parallel_backends.py:
```diff
@@ -260,23 +305,23 @@ def _read_stdout(self, fd, mask):
             raise ValueError("Incorrect padding for data length %d bytes" %
-                             len(data_bytes), "(mod 4 = %d)" %
-                             len(data_bytes) % 4)
+                             len(data_bytes) + " (mod 4 = %d)" %
```
Your right that there shouldn't be a need for this, as I wouldn't expect it to hit this anymore. I think the UserWarning gives enough debugging information (actual data length and expected length) that this can be removed.

blakecaldwell replied:
> In the future do you think it will be necessary to add print statements indicating time passing/progress while data is transferred? As we add the ability to record from many cells simultaneously, I can see the wait times getting pretty large.

That seems like a good idea to me.

blakecaldwell replied:
> > In the future do you think it will be necessary to add print statements indicating time passing/progress while data is transferred? As we add the ability to record from many cells simultaneously, I can see the wait times getting pretty large.
> 
> That seems like a good idea to me.

I'm going to punt this to later if/when we see that the delay is unexpectedly long.

blakecaldwell commented:
> Looks very promising! Code gets better and better with each PR.
> 
> So, I did not get through the entire PR. I still need to do `test_mpi_backends.py`. Please interpret all the flurry of comments as my excitement to see interesting progress, specially with the testing procedure!
> 
Thanks for reviewing the code. I'm in a very busy period for the next two weeks, so I probably won't be able to implement significant code changes in the meantime. Also if I don't reply to comments for a bit, that's why :)

> One last question, wondering how all this compares to [Popen.communicate](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate)? If there are system library alternatives, we should try to use them.

I used `Popen.communicate()` before #185. That summarizes one reason. Additionally, the issues seen recently with interpreting data/error messages cannot be resolved with `Popen.communicate()` because it blocks until the process completes.

blakecaldwell commented on hnn_core/tests/test_parallel_backends.py:
```diff
@@ -13,23 +14,29 @@
+    return dpls, net
 
+
+def run_hnn_core_reduced(backend=None, n_procs=None, n_jobs=1):
```
That works.

blakecaldwell commented on hnn_core/tests/test_parallel_backends.py:
```diff
@@ -3,6 +3,7 @@
 import pytest
 import io
 from contextlib import redirect_stdout
+from multiprocessing import cpu_count
```
Sure

blakecaldwell commented on hnn_core/tests/test_parallel_backends.py:
```diff
@@ -79,65 +93,94 @@ def run_hnn_core(backend=None, n_jobs=1):
+
+    def test_run_mpibackend(self):
+        """Test running a MPIBackend on reduced model"""
+        global dpls_reduced_default, dpls_reduced_mpi
```
What specific part of this block are you suggesting using fixtures for? I looked at fixtures and they didn't handle the dependencies they way I wanted. There wasn't one check that I wanted to run over and over, but a series of tests. Note that `test_mpi_nprocs` doesn't run a simulation, but would be important if someone tries running the tests on a VM with 1 processor.

You could use fixtures with parameters to replace `run_hnn_core_reduced`, (similar to `pytest.mark.parametrize` mentioned below), but tbh it just looks like adding complexity to a very simple thing.

blakecaldwell commented on hnn_core/tests/test_parallel_backends.py:
```diff
@@ -79,65 +93,94 @@ def run_hnn_core(backend=None, n_jobs=1):
+        oversubscribed = cpu_count() * 1
+        run_hnn_core_reduced(backend='mpi', n_procs=oversubscribed)
 
+    def test_compare_hnn_core(self):
```
Hmm. Could you please explain the rationale for creating a complicated function that all the tests call with different parameters? I intentionally broke up the function that was coded before because I thought it was convoluted (when are full vs. reduced trials being run, what data output is being compared, etc.) and would be hard to debug.

jasmainak replied:
Hard to explain ... but I pushed a commit 077f87cd to show what I meant. It's a bit simpler no? For debugging I always add ``--pdb`` flag at the end and you know what parameters were passed to the function.

blakecaldwell replied:
Okay, I was thinking of different parameters (i.e. `compare=True`). The way you coded it seems simpler. Thanks for demonstrating.

jasmainak commented on hnn_core/tests/test_mpi_child.py:
```diff
@@ -56,10 +131,19 @@ def test_empty_data():
+    with pytest.warns(UserWarning) as record:
+        backend._process_child_data(pickled_bytes, expected_len)
+
+    expected_string = "Length of received data unexpected. " + \
+        "Expecting %d bytes, got %d" % (expected_len, len(pickled_bytes))
+
+    assert len(record) == 1
+    assert record[0].message.args[0] == expected_string
```
use the `match` argument? I guess it's enough to check if the first part of the string matches?

blakecaldwell replied:
Sure, but why use a function if the operator does the same thing?

blakecaldwell replied:
I thought there was value in checking the entire string matched (I see your point if you only check the first part).

However, just like in the comment below, I don't think the test is necessary anymore!

blakecaldwell replied:
I'm leaving this in as the full string because we are checking that the mismatch between received and expected bytes. Unfortunately, with base64 this should stay in. If someone needs to debug it, it will be useful to know that there was a mismatch.

jasmainak commented on hnn_core/tests/test_mpi_child.py:
```diff
@@ -56,10 +131,19 @@ def test_empty_data():
+    with MPISimulation(skip_mpi_import=True) as mpi_sim:
+        pickled_bytes = mpi_sim._pickle_data({})
+
+    expected_len = len(pickled_bytes) + 1
```
I didn't notice it before but why the + 1 ?

blakecaldwell replied:
this triggered the UserWarning, but it may not be necessary when base64 encoding is removed from this PR

jasmainak reviewed (commented):
Okay this was last major review I'd say. More or less understand the idea behind the code now. Thanks for taking a stab at fixing it!

blakecaldwell commented on hnn_core/parallel_backends.py:
```diff
@@ -236,22 +249,54 @@ def __exit__(self, type, value, traceback):
-        # data is signalled by the process terminating
         return None
 
     def _read_stdout(self, fd, mask):
```
No, but it's part of the `select()` callback API. Its purpose is to check if the callback is triggered by a read or a write. We only have one-way communication in our case.

blakecaldwell commented:
Hi all, starting my update to this PR today.

blakecaldwell commented:
@jasmainak @ntolley I think that I've addressed all comments above and this PR is ready once it goes green. Thanks for reviewing... hope the code is reasonable to follow now.

jasmainak commented:
PR is good enough to merge. Let's move on and make improvements in future PRs, I realize your time is limited @blakecaldwell . Thanks for the efforts so far!

jasmainak commented:
@blakecaldwell is the "notes for contributors" section in this document still up-to-date: https://github.com/jonescompneurolab/hnn-core/blob/master/doc/parallel.rst ? A quick PR to update it would be very much appreciated!
